2025022700 OS crash when location globally disabled

I have previously sent an email to logs@grapheneos.org regarding a similar issue with log files of some apps with the issue. This time, the app causes the phone to crash and reboot.


```
type: crash

SystemUptimeMs: 1765209
Process: system_server
Timestamp: 2025-02-28 01:55:01.063+0800
Build: google/oriole/oriole:15/AP4A.250205.002/2025022700:user/release-keys
Crash-Handler: com.android.internal.os.RuntimeInit$KillApplicationHandler
Loading-Progress: 1.0
Dropped-Count: 0

java.lang.AssertionError: java.lang.IllegalArgumentException: unregistered listener cannot be flushed
	at android.location.provider.LocationProviderBase$Service.lambda$flush$1(LocationProviderBase.java:349)
	at android.location.provider.LocationProviderBase$Service$$ExternalSyntheticLambda3.run(D8$$SyntheticClass:0)
	at android.os.Handler.handleCallback(Handler.java:991)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loopOnce(Looper.java:232)
	at android.os.Looper.loop(Looper.java:317)
	at com.android.server.SystemServer.run(SystemServer.java:995)
	at com.android.server.SystemServer.main(SystemServer.java:673)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:591)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:865)
Caused by: java.lang.IllegalArgumentException: unregistered listener cannot be flushed
	at com.android.internal.util.Preconditions.checkArgument(Preconditions.java:64)
	at android.location.LocationManager.requestFlush(LocationManager.java:1734)
	at com.android.location.fused.FusedLocationProvider$ChildLocationListener.flush(FusedLocationProvider.java:314)
	at com.android.location.fused.FusedLocationProvider.onFlush(FusedLocationProvider.java:145)
	at android.location.provider.LocationProviderBase$Service.flush(LocationProviderBase.java:344)
	at com.android.server.location.provider.proxy.ProxyLocationProvider$2.run(ProxyLocationProvider.java:207)
	at com.android.server.servicewatcher.ServiceWatcherImpl$MyServiceConnection.runOnBinder(ServiceWatcherImpl.java:257)
	at com.android.server.servicewatcher.ServiceWatcherImpl.lambda$runOnBinder$0(ServiceWatcherImpl.java:126)
	at com.android.server.servicewatcher.ServiceWatcherImpl.$r8$lambda$GCGDYqQZRK7jCQUNZnY8QjXgR2Y(ServiceWatcherImpl.java:0)
	at com.android.server.servicewatcher.ServiceWatcherImpl$$ExternalSyntheticLambda1.run(R8$$SyntheticClass:0)
	at android.os.Handler.handleCallback(Handler.java:991)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loopOnce(Looper.java:232)
	at android.os.Looper.loop(Looper.java:317)
	at android.os.HandlerThread.run(HandlerThread.java:85)
	at com.android.server.ServiceThread.run(ServiceThread.java:46)
```

Can confirm this issue on Pixel 6a. 

```

type: crash

SystemUptimeMs: 3020889
Process: system_server
Timestamp: 2025-02-27 21:56:08.932+0100
Build: google/bluejay/bluejay:15/AP4A.250205.002/2025022700:user/release-keys
Crash-Handler: com.android.internal.os.RuntimeInit$KillApplicationHandler
Loading-Progress: 1.0
Dropped-Count: 0

java.lang.AssertionError: java.lang.IllegalArgumentException: unregistered listener cannot be flushed
	at android.location.provider.LocationProviderBase$Service.lambda$flush$1(LocationProviderBase.java:349)
	at android.location.provider.LocationProviderBase$Service$$ExternalSyntheticLambda3.run(D8$$SyntheticClass:0)
	at android.os.Handler.handleCallback(Handler.java:991)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loopOnce(Looper.java:232)
	at android.os.Looper.loop(Looper.java:317)
	at com.android.server.SystemServer.run(SystemServer.java:995)
	at com.android.server.SystemServer.main(SystemServer.java:673)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:591)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:865)
Caused by: java.lang.IllegalArgumentException: unregistered listener cannot be flushed
	at com.android.internal.util.Preconditions.checkArgument(Preconditions.java:64)
	at android.location.LocationManager.requestFlush(LocationManager.java:1734)
	at com.android.location.fused.FusedLocationProvider$ChildLocationListener.flush(FusedLocationProvider.java:314)
	at com.android.location.fused.FusedLocationProvider.onFlush(FusedLocationProvider.java:145)
	at android.location.provider.LocationProviderBase$Service.flush(LocationProviderBase.java:344)
	at com.android.server.location.provider.proxy.ProxyLocationProvider$2.run(ProxyLocationProvider.java:207)
	at com.android.server.servicewatcher.ServiceWatcherImpl$MyServiceConnection.runOnBinder(ServiceWatcherImpl.java:257)
	at com.android.server.servicewatcher.ServiceWatcherImpl.lambda$runOnBinder$0(ServiceWatcherImpl.java:126)
	at com.android.server.servicewatcher.ServiceWatcherImpl.$r8$lambda$GCGDYqQZRK7jCQUNZnY8QjXgR2Y(ServiceWatcherImpl.java:0)
	at com.android.server.servicewatcher.ServiceWatcherImpl$$ExternalSyntheticLambda1.run(R8$$SyntheticClass:0)
	at android.os.Handler.handleCallback(Handler.java:991)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loopOnce(Looper.java:232)
	at android.os.Looper.loop(Looper.java:317)
	at android.os.HandlerThread.run(HandlerThread.java:85)
	at com.android.server.ServiceThread.run(ServiceThread.java:46)
```

--- a/packages/FusedLocation/src/com/android/location/fused/FusedLocationProvider.java
+++ b/packages/FusedLocation/src/com/android/location/fused/FusedLocationProvider.java
@@ -128,10 +128,12 @@ public class FusedLocationProvider extends LocationProviderBase {
     public void onFlush(OnFlushCompleteCallback callback) {
         synchronized (mLock) {
             AtomicInteger flushCount = new AtomicInteger(0);
-            if (mGpsPresent) {
+            boolean flushGps = mGpsListener.isActive();
+            if (flushGps) {
                 flushCount.incrementAndGet();
             }
-            if (mNlpPresent) {
+            boolean flushNetwork = mNetworkListener.isActive();
+            if (flushNetwork) {
                 flushCount.incrementAndGet();
             }
 
@@ -141,10 +143,10 @@ public class FusedLocationProvider extends LocationProviderBase {
                 }
             };
 
-            if (mGpsPresent) {
+            if (flushGps) {
                 mGpsListener.flush(wrapper);
             }
-            if (mNlpPresent) {
+            if (flushNetwork) {
                 mNetworkListener.flush(wrapper);
             }
         }
@@ -278,6 +280,10 @@ public class FusedLocationProvider extends LocationProviderBase {
             synchronized (mLock) {
                 mLocation = null;
             }
+        }
+
+        boolean isActive() {
+            return getInterval() != INTERVAL_DISABLED;
         }
 
         private void resetProviderRequest(long newInterval) {

